lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -362,10 +409,41 @@ private TaskStateEvent GetTaskErrorEventByExceptionType(IFailedTask failedTask)
             {
                 return TaskStateEvent.FailedTaskCommunicationError;
             }
-            else
+            if (exception is IMRUTaskSystemException)
             {
                 return TaskStateEvent.FailedTaskSystemError;
             }
+
+            // special case for communication error during group communication initialization
+            if (exception is TaskClientCodeException)
+            {
+                {
+                    // try extract cause and check whether it is InjectionException for GroupCommClient
+                    if (exception.InnerException != null &&
+                        exception.InnerException is InjectionException &&
+                        exception.InnerException.Message.Contains("GroupCommClient"))
+                    {
+                        Logger.Log(Level.Info, "GetTaskErrorEventByExceptionType:FailedTaskCommunicationError with task id {0}", failedTask.Id);
+                        return TaskStateEvent.FailedTaskCommunicationError;
+                    }
+                }
+            }
+
+            if (exception is AggregateException)
+            {
+                {
+                    // try extract cause and check whether it is InjectionException for GroupCommClient
+                    if (exception.InnerException != null &&
+                        exception.InnerException is IOException)
+                    {
+                        Logger.Log(Level.Info, "GetTaskErrorEventByExceptionType:AggregateException-IOException with task id {0}", failedTask.Id);
+                        return TaskStateEvent.FailedTaskCommunicationError;
+                    }
+                }
+            }
```

Comment:
To me this whole function is extremely complex and there are no guarantees that we still have covered all the cases. Is it that certain core REEF parts require some changes/improvements and we are introducing complexity here because of that.